### PR TITLE
implement `date` and `time` conversions (Fixes #445)

### LINKED ISSUE #445: mongodb: implement `date` and `time` syntax
The parser supports SQL standard syntax for date and time values (as in `select date '2014-11-17', time '12:00:00'`), but MongoDB provides no convenient way of representing these.

We'll need to emulate them using date/time operations.

See #442.

### PR DESCRIPTION
**Pull request by mossprescott**
`Date` and `Time`, as well as `Timestamp` and `Interval`, all become
first-class values in `Data`, so that the front-end can communicate the
semantics unambiguously in the logical plan.

`Date` comparisons get special handling in the planner, since only
timestamp-equivalent values actually exist in MongoDB, but the assumption is
that people are using them to store dates in some cases, and comparing them
involves computing the end-of-day timestamp.

`Time` values, on the other hand, become simple strings (which happen to be
comparable as is). This meant producing a rather complicated JS expression, so
our `whenDefined` scheme is significantly beefed up. For example, we now check
for the presence of a method before calling it (with a compound test as in
`(value != null && value.getUTCSeconds != null) ? value.getUTCSeconds() ?
undefined`), but do *not* do that test in cases where it's clear that we know
what we're doing (as in `['CO', 'CA'].indexOf(state)`).

The operators in `MathLib` now accept temporal values where the result is
sensible, which flushed out quite a few bugs and corner cases in those type
checkers. There is probably more to do here, but the current implementation
supports several useful ways of combining constants and values.

Along the way, fixed some of the worst cases associated with #571, but there
are still likely many combinations of constants and mixed types that don't
work right.

### CODE REVIEW

**Comment by jdegoes** [comment]
I reviewed 24,000 lines of code today so maybe I'll let @sellout take this one. ;-)

File: core/src/main/scala/slamdata/engine/std/math.scala [hunk 1]
```diff
@@ -66,12 +73,16 @@ trait MathLib extends Library {
...
    */
-  val Divide = Mapping("(/)", "Divides one numeric value by another (non-zero) numeric value", NumericDomain,
+  val Divide = Mapping("(/)", "Divides one numeric or interval value by another (non-zero) numeric value", (Type.Numeric | Type.Interval) :: (Type.Numeric | Type.Interval) :: Nil,
     (partialTyperV {
       case v1 :: Type.Const(Data.Number(v2)) :: Nil if (v2.doubleValue == 1.0) => success(v1)
       case v1 :: Type.Const(Data.Number(v2)) :: Nil if (v2.doubleValue == 0.0) => failure(NonEmptyList(GenericError("Division by zero")))
       case Type.Const(Data.Int(v1)) :: Type.Const(Data.Int(v2)) :: Nil => success(Type.Const(Data.Int(v1 / v2)))
       case Type.Const(Data.Number(v1)) :: Type.Const(Data.Number(v2)) :: Nil => success(Type.Const(Data.Dec(v1 / v2)))
+
+      // TODO: handle interval divided by Dec (not provided by threeten)
```
Comment (sellout) [inline core/src/main/scala/slamdata/engine/std/math.scala]: Can you create an issue for this and add the issue number to this comment?

File: it/src/test/resources/tests/time.test [hunk 2]
```diff
@@ -0,0 +1,15 @@
...
+  "query": "select \"day\", time_of_day(\"date\") as tod, time_of_day(\"day\") as \"not a date\", time_of_day(missing) as missing from days
+    where time_of_day(\"date\") >= time '08:00' and time_of_day(\"date\") < time '10:20:30.400'",
+  
+  "predicate": "containsExactly",
+  "expected": [
+    { "value": { "day": "Tuesday",   "tod": "08:00:00.000", "not a date": null, "missing": null } },
+    { "value": { "day": "Wednesday", "tod": "09:00:00.000", "not a date": null, "missing": null } },
+    { "value": { "day": "Thursday",  "tod": "10:00:00.000", "not a date": null, "missing": null } }
+  ]
+}
```
Comment (sellout) [inline it/src/test/resources/tests/time.test]: Your new files are missing final newlines.

Also ↓

**Comment by sellout** [comment]
*phew* that was a slog. Looks good, though – just a couple super-minor comments.

File: it/src/test/resources/tests/time.test (hunk 2 above)
Comment (jdegoes) [inline it/src/test/resources/tests/time.test]: Also, it looks like only 1 integration test was added. I think you should get another 2 or so in here with all these changes. Our coverage for date / time functionality has not been great.

**Pull request merged**